src/roles.c

```diff
@@ -0,0 +1,525 @@
+
+/* Overview
+ * --------
+ *
+ * This file implements automatic role management for dqlite servers. When
+ * automatic role management is enabled, servers in a dqlite cluster will
+ * autonomously (without client intervention) promote and demote each other
+ * to maintain a specified number of voters and standbys, taking into account
+ * the health, failure domain, and weight of each server.
+ *
+ * We implement two ingredients of role management: adjustments and handovers.
+ * Adjustment runs on the cluster leader every tick (the frequency is defined
+ * in server.c). The first step is to "poll" every server in the cluster to find
+ * out whether it's online, and if so, its failure domain and weight. It demotes
+ * to spare any servers that appear to have gone offline, then, if the numbers
+ * of (online) voters and standbys don't match the target values, chooses
+ * servers that should be promoted or demoted. The preference ordering for
+ * promotion is based on the failure domains and weights previously gathered,
+ * and is defined in compareNodesForPromotion, below.
+ *
+ * The actual roles changes are computed in a batch each time adjustment
+ * occurs, and are stored in a queue. Individual "change records" are taken
+ * off this queue and applied asynchronously. Since we only have a blocking
+ * client implementation available, the exchanges of requests and responses
+ * that implements polling a single server happens on the libuv blocking
+ * thread pool (see pollClusterAfterWorkCb). We don't start a new round of
+ * adjustment if a "tick" occurs while the queue of changes from the last
+ * round is still nonempty.
+ *
+ * A handover is triggered when we call dqlite_node_handover on a node that's
+ * the current cluster leader, or is a voter. Before shutting down for real,
+ * the node in question tries to cause another node to become leader (using
+ * raft_transfer), if applicable, and then promotes another node to voter
+ * (if possible) before demoting itself. This is intended to smooth over
+ * availability problems that can result if a privileged node (leader or
+ * non-leader voter) crashes out of the cluster unceremoniously. The handover
+ * task also needs to poll the cluster to figure out which nodes are good
+ * candidates for promotion to voter.
+ *
+ * Unresolved
+ * ----------
+ *
+ * - Should the failure-domains accounting for standbys use information about
+ *   voters' failure domains? Vice versa?
+ * - Should we try multiple candidates when doing an adjustment, if the
+ *   preferred candidate can't be promoted?
+ * - Should we retry when some step in the handover process fails? How, and
+ *   how many times?
+ * - Should we have dedicated code somewhere to (possibly) promote newly-
+ *   joined nodes? go-dqlite does this, but I'm not convinced it's important,
+ *   or that it should run on the server if we do decide we want it.
+ */
+
+/* XXX */
+#define NUM_TRACKED_DOMAINS 5
+
+struct change_record
+{
+	uint64_t id;
```

Comment:
I think we had this conversation before, the principle is that model/language types are independent from protocol/serialization concerns.

At some point we might change the serialization type of a given model/language type without changing the model/language type (thus without breaking the API/ABI). For example we might change the protocol to allow for greater values, without requiring a language/model change.

One recent example of that was the allowed number of statement parameters. The protocol version X only allowed for 256 parameters. But the associated language type was **not** `uint8_t` (or whatever Go equivalent). By separating the two concerns we were able to bump the protocol version to version X+1 with a different serialization format/type for the number of the statement parameters without needing to change the language. In the implementation there was a missing check for the 256 limit, but that's fine, it was just a bug, and the fix was not to change the language type from `int` to `uint8_t` (or whatever exact Go types), but rather to add the check. 

In this specific case the current situation is that the failure domain must be a value between `0` and `2^64`, but we don't want to enforce that with a `uint64_t` type, we want to enforce that with a check in the code (and add documentation if missing). This way we can evolve the wire protocol in the future, if required. We're most probably not going to need that for the failure domain values specifically, but this the general idea.

> Is our code actually prepared to deal with a hypothetical target where `unsigned long long` and `uint64_t` are not the same type?

That I didn't check. If it's not, it should be fixed with a check on the value, and documented.

Note that boundary checks are just one type of model value checks. In general any expectation we have on a value (e.g. the value should be "odd", or it should be a multiple of 3 or whatever) should be explicitly coded.

> It seems to me that if the IDs, failure domains, etc. are required by the wire protocol to be 64 bits, we should just be using `uint64_t` or a typedef of it everywhere.

As said, the wire protocol is opaque, and can change. What should be more stable is the language/model API.

> Using the "wobbly-width" C integer types in that kind of situation is (IMO) either pointless, if the width of the type is effectively fixed (as for `unsigned long long`), or buggy, if the width can vary across targets.

This is indeed the point, the width of the type is **not** effectively fixed **by design**. The current boundary values are [0..2^64] (and as said they should be documented and checked), but those could change in the future without changing the **type**.

Reply:
> (I don't find the argument from optimization persuasive because I can't think of a target where `unsigned long long` is 128 bits, and even on such a target the gains from using a 128-bit type over a 64-bit type seem to me very unlikely to justify the complexity incurred by mixing `unsigned long long` and `uint64_t` when these are not the same width.)

The optimization argument is ancillary and definitely not the main, the point is more to separate concerns conceptually. As said, failure domains don't **need** to be 64 bits, and should not be modeled as such. As said, this doesn't matter much in this specific case, but it might matter in other cases, and for consistency we apply the same approach everywhere.

Reply:
In a 32-bit world, it would be good practice to use `int` for your APIs, not `int32_t`. At that point you might have checks in place to ensure that your values were at most 2^32 (because for example you were serializing them using `int32_t`). Then you start supporting new 64-bit architectures, and you don't need to change the API (`int` to `int64_t`), but you might change the serialization and relax your boundary checks to allow for greater values. User code does not need to change.